vault/ha_test.go

```diff
@@ -85,3 +88,50 @@ func TestGrabLockOrStop(t *testing.T) {
 	}
 	workerWg.Wait()
 }
+
+func TestCoreRestart(t *testing.T) {
+	t.Parallel()
+
+	t.Run("active", func(t *testing.T) {
+		t.Skip("there is a data-race in waitForLeadership: https://github.com/openbao/openbao/blob/5a93ec0549a88516a8a15e96ef74dadac8ed506f/vault/ha.go#L676-L687")
+		testCoreRestart(t, 0)
+	})
```

Comment:
For now, the restart test is skipped because of  a data-race in the `waitForLeadership` implementation https://github.com/openbao/openbao/blob/5a93ec0549a88516a8a15e96ef74dadac8ed506f/vault/ha.go#L676-L687

If `l.lockOrStop()` returns `false` the code still modifies the state. I tried to fix this, by changing it to always lock and return in the end, if the context was cancelled. But other parts of the shutdown logic seem to rely on the unlocked state and it caused a dead-lock. So fixing this would be a bigger change.
As for now `core.restart()` is only called on standby nodes and the race only happens on an active node this PR does not make the situation worse. Therefore I decided to fix this in a separate PR (targeting main)